test/src/test/java/org/corfudb/integration/ServerRestartIT.java

```diff
@@ -462,6 +468,213 @@ public void sequencerTailsRecoveryTest() throws Exception {
         assertThat(shutdownCorfuServer(corfuServerProcess)).isTrue();
 
     }
+
+    /**
+     * If a tokenResponse is received in the previous epoch,
+     * a write should discard the tokenResponse and throw an exception.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void discardTokenReceivedInPreviousEpoch() throws Exception {
+        final int timeToWaitInSeconds = 3;
+
+        Process corfuServerProcess = runCorfuServer();
+        final CorfuRuntime corfuRuntime = createDefaultRuntime();
+
+        // wait for this server long enough to start (by requesting token service)
+        TokenResponse firsttr = corfuRuntime.getSequencerView().nextToken(Collections.emptySet(),
+                1);
+
+        assertThat(shutdownCorfuServer(corfuServerProcess)).isTrue();
+
+        corfuServerProcess = runCorfuServer();
+
+        corfuRuntime.invalidateLayout();
+        TokenResponse tr = corfuRuntime.getSequencerView().nextToken(Collections.emptySet(), 1);
+
+        assertThat(tr.getEpoch())
+                .isEqualTo(1);
+
+        // Force the token response to have epoch = 0, to simulate a request received in previous epoch
+        TokenResponse mockTr = new TokenResponse(tr.getToken().getTokenValue(), tr.getEpoch() - 1, Collections.emptyMap());
+
+        byte[] testPayload = "hello world".getBytes();
+
+        // Should succeed. internally, it will refresh the token.
+        CompletableFuture cf = CFUtils.within(CompletableFuture.supplyAsync(() -> {
+            corfuRuntime.getAddressSpaceView().write(mockTr, testPayload);
+            return true;
+        }), Duration.ofSeconds(timeToWaitInSeconds));
+
+        try {
+            cf.get();
+        } catch (Exception e) {
+            assertThat(e.getCause()).isNotInstanceOf(TimeoutException.class);
+            assertThat(e.getCause()).isInstanceOf(StaleTokenException.class);
+        }
+
+        assertThat(shutdownCorfuServer(corfuServerProcess)).isTrue();
+    }
+
+    /**
+     * loop that randomizes a combination of the following activities:
+     * (i) map put()s
+     * (ii) checkpoint/trim
+     * (iii) leaving holes behind
+     * (iv) server shutdown/restart
+     *
+     * @throws Exception
+     */
+
+  public void sequencerTailsRecoveryLoopTest() throws Exception {
```

Comment:
I think this is a great test but we should break the long method into smaller methods and have more tests with different combinations of those smaller methods. And some tests can vary on the number of iterations we run.